### DISCUSSION TITLE
Get performance estimates without deployment

### DISCUSSION THREAD

**Discussion by hossein1387** [Xilinx/finn#428] [Q&A]
Hi,

I am trying to get some performance estimate of running quantized models on FINN. What I want to do is given an onnx model (a classification model), what would be the throughput of it using FINN? I cam across these code across finn repo: https://github.com/Xilinx/finn-base/blob/52ce94b9f39179189dd44179c460d0b145cafcf1/src/finn/core/throughput_test.py#L115
https://github.com/Xilinx/finn/blob/d1cc9cf94f1c33354cc169c5a6517314d0e94e3b/tests/fpgadataflow/test_fpgadataflow_globalaccpool.py#L127
https://github.com/daiki98/finn/blob/b8f1635f557891f60ede71a54ee1f4385f5e1c6a/tests/fpgadataflow/test_fpgadataflow_channelwise_ops.py#L159
However, I am not able to run it. Is there any sample code that I can follow?

**Comment by rpitonak** [comment]
I do not know if it is helpful but you can try to use `AllocateResources` transformation from FINN experimental.

https://github.com/Xilinx/finn-experimental/blob/main/src/finn/transformation/fpgadataflow/allocate_resources.py

It utilizes `SetFolding` transformation

https://github.com/Xilinx/finn/blob/main/src/finn/transformation/fpgadataflow/set_folding.py 

which according to docstring:
> In the returned model, each node's
    cycles_estimate attribute will be set to its estimated number of cycles
    
Example code:
```python
from finn.transformation.fpgadataflow.allocate_resources import AllocateResources
from finn.core.modelwrapper import ModelWrapper

model = ModelWrapper(f"{MODEL_PATH}-dataflow_model.onnx")

model = model.transform(AllocateResources(clk_ns=10, fps_target=20, platform="Pynq-Z1"))

model.save(f"{MODEL_PATH}-dataflow_model-folded.onnx")
```

Then you can inspect the resulting `.onnx` model using netron or with the code similar to this:

```python
model = ModelWrapper(f"{MODEL_PATH}-dataflow_model-folded.onnx")

fc_layers = model.get_nodes_by_op_type("StreamingFCLayer_Batch")
for fcl in fc_layers:
    for attr in fcl.attribute:
        if attr.name == "cycles_estimate":
            print(f"{fcl.name} {attr.name} {attr.i}")
```

The output will be something like this:
```
StreamingFCLayer_Batch_0 cycles_estimate 27648
StreamingFCLayer_Batch_1 cycles_estimate 20736
StreamingFCLayer_Batch_2 cycles_estimate 524288
...
```

I am not sure if this is the best possible solution. Let me know if something is unclear.

**Reply by hossein1387** [reply]
Thanks for the reply, I tried my own model with the example code you provided and I am getting this error:

```
finn/src/finn/analysis/fpgadataflow/dataflow_performance.py", line 75, in dataflow_performance
    critical_path_cycles = max(latency_at_node_output.values())
ValueError: max() arg is an empty sequence
    critical_path_cycles = max(latency_at_node_output.values())
ValueError: max() arg is an empty sequence

```

And when I debug the line it shows, I can see that none of the nodes in my model are an `fpgadataflow` node, hence making the `dataflow_performance` return an empty list. So my question is, how to transform my model to an `fpgadataflow` model?

**Reply by rpitonak** [reply]
Hello,

You need to have your model consisting of HLS layers before executing this code. For example, if we consider the example:
https://github.com/Xilinx/finn/blob/main/notebooks/end2end_example/bnn-pynq/cnv_end2end_example.ipynb

You should call this code after line 

```python
dataflow_model.save(build_dir + "/end2end_cnv_w1a1_dataflow_model.onnx")
```

when the model is ready for synthesis. In this case it would be:

```python
from finn.transformation.fpgadataflow.allocate_resources import AllocateResources
from finn.core.modelwrapper import ModelWrapper

model = ModelWrapper(build_dir + "/end2end_cnv_w1a1_dataflow_model.onnx")

model = model.transform(AllocateResources(clk_ns=10, fps_target=20, platform="Pynq-Z1"))

model.save((build_dir + "/end2end_cnv_w1a1_dataflow_model-folded.onnx")
```

Let me know if something is unclear.

**Reply by hossein1387** [reply]
Thanks a lot for your reply. I am getting errors regarding `FIN_BUILD_DIR` and after digging deeper, I realized that most of the tutorials I have seen rely on a docker image and xilinx tools. I have a Mac and I can install the Vivado Webpack on docker image. I was wondering does this setup satisfies the requirement to run the code above? Is there any way to run the profiler code you have without access to Xilinx tools?

**Reply by rpitonak** [reply]
I do not think that this is possible without the Xilinx tools (Vivado especially). Somebody correct me if I am wrong.

Running FINN inside docker image with Vivado installed on the host machine is definitely what I would recommend if you want to experiment with the FINN framework.

**Reply by hossein1387** [reply]
Thanks for your reply. I did what you suggest. I am now using a Host machine (not Mac) with Vivado installed on it. I followed all the instructions available here: https://finn.readthedocs.io/en/latest/getting_started.html
I am now trying to use `build_dataflow` command to convert my model. For that I followed the instructions ( I have a folder in my finn directory called `dataflow_build_dir` and inside it I have `dataflow_build_config.json` and   `model.onnx`. I then use the following command:

`sudo ./run-docker.sh build_dataflow ./dataflow_build_dir`

Here is the output log: [output.log](https://github.com/Xilinx/finn/files/7551794/output.log)

and when I don't use `sudo` (I already added my user to docker group) I get same error plus the following:

```
Successfully built ca037cd61279
Successfully tagged xilinx/finn:v0.7-dirty.xrt_202010.2.7.766_18.04-amd64-xrt
docker run -t --rm -it --tty --init --hostname finn_dev_askari -e SHELL=/bin/bash -v /home/askari/MyRepos/finn:/workspace/finn -v /tmp/finn_dev_askari:/tmp/finn_dev_askari -e FINN_BUILD_DIR=/tmp/finn_dev_askari -e FINN_ROOT=/workspace/finn -e LOCALHOST_URL=localhost -e VIVADO_IP_CACHE=/tmp/finn_dev_askari/vivado_ip_cache -e PYNQ_BOARD=Pynq-Z1 -e PYNQ_IP= -e PYNQ_USERNAME=xilinx -e PYNQ_PASSWORD=xilinx -e PYNQ_TARGET_DIR=/home/xilinx/finn_dev_askari -e NUM_DEFAULT_WORKERS=4 -v /etc/group:/etc/group:ro -v /etc/passwd:/etc/passwd:ro -v /etc/shadow:/etc/shadow:ro -v /etc/sudoers.d:/etc/sudoers.d:ro -v /home/askari/MyRepos/finn/ssh_keys:/home/askari/.ssh --user 1005:999 -v /opt/Xilinx:/opt/Xilinx -e XILINX_VIVADO=/opt/Xilinx/Vivado/2021.2 -e VIVADO_PATH=/opt/Xilinx/Vivado/2021.2 -v /home/askari/MyRepos/finn/dataflow_build_dir:/home/askari/MyRepos/finn/dataflow_build_dir xilinx/finn:v0.7-dirty.xrt_202010.2.7.766_18.04-amd64-xrt build_dataflow /home/askari/MyRepos/finn/dataflow_build_dir
Obtaining file:///workspace/finn
Installing collected packages: finn
  Running setup.py develop for finn
    ERROR: Command errored out with exit status 1:
     command: /opt/conda/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/workspace/finn/setup.py'"'"'; __file__='"'"'/workspace/finn/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps --user --prefix=
         cwd: /workspace/finn/
    Complete output (3 lines):
    running develop
    running egg_info
    error: [Errno 13] Permission denied
    ----------------------------------------
ERROR: Command errored out with exit status 1: /opt/conda/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/workspace/finn/setup.py'"'"'; __file__='"'"'/workspace/finn/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps --user --prefix= Check the logs for full command output.
WARNING: Unable to find /settings64.sh
WARNING: Functionality dependent on Vitis will not be available.
WARNING: If you need Vitis, ensure VITIS_PATH is set correctly and mounted into the Docker container.
Found Vivado at /opt/Xilinx/Vivado/2021.2
/usr/local/bin/finn_entrypoint.sh: line 94: exec: build_dataflow: not found
```

The main error is:

`/usr/local/bin/finn_entrypoint.sh: line 94: exec: build_dataflow: not found`

I am not sure what is wrong with my setup. Any suggestion?

**Reply by hossein1387** [reply]
I added ENV `VITIS_PATH="/opt/Xilinx/Vitis_HLS/2021.2"` to `Dockerfile.finn` and got a bit closer but still getting error:
```
Repos/finn/dataflow_build_dir
Obtaining file:///workspace/finn
Installing collected packages: finn
  Running setup.py develop for finn
Successfully installed finn
Found Vitis at /opt/Xilinx/Vitis_HLS/2021.2
XILINX_XRT      : /opt/xilinx/xrt
PATH            : /opt/xilinx/xrt/bin:/opt/Xilinx/Vitis_HLS/2021.2/bin:/opt/Xilinx/Model_Composer/2021.2/bin:/opt/Xilinx/DocNav:/opt/Xilinx/Vivado/2021.2/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/workspace/oh-my-xilinx
LD_LIBRARY_PATH : /opt/xilinx/xrt/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
PYTHONPATH     : /opt/xilinx/xrt/python:
Found XRT at /opt/xilinx/xrt
/usr/local/bin/finn_entrypoint.sh: line 94: exec: build_dataflow: not found
```

**Reply by hossein1387** [reply]
I also tried `3-build-accelerator-with-finn.ipynb` note book with bunch of your own models and none of them work. My last attempt was with this model: 
`https://github.com/Xilinx/finn-examples/releases/download/v0.0.1a/onnx-models-resnet50.zip`

and after about 10 min I got this error:

```
Building dataflow accelerator from resnet50_w1a2_exported.onnx
Intermediate outputs will be generated in $(/home/askari/finn_output)
Final outputs will be generated in output_estimates_only
Build log is at output_estimates_only/build_dataflow.log
Running step: step_qonnx_to_finn [1/8]
/home/askari/MyRepos/finn-base/src/finn/core/modelwrapper.py:91: UserWarning: Some old-style domain attributes were automatically converted to new-style,
                i.e. domain=finn to domain=finn.custom_op.<general|fpgadataflow|...>
  i.e. domain=finn to domain=finn.custom_op.<general|fpgadataflow|...>"""
Running step: step_tidy_up [2/8]
Running step: step_streamline [3/8]
Traceback (most recent call last):
  File "/home/askari/MyRepos/finn/src/finn/builder/build_dataflow.py", line 166, in build_dataflow_cfg
    model = transform_step(model, cfg)
  File "/home/askari/MyRepos/finn/src/finn/builder/build_dataflow_steps.py", line 263, in step_streamline
    model = model.transform(Streamline())
  File "/home/askari/MyRepos/finn-base/src/finn/core/modelwrapper.py", line 142, in transform
    transformed_model
  File "/home/askari/MyRepos/finn/src/finn/transformation/streamline/__init__.py", line 97, in apply
    model = model.transform(trn)
  File "/home/askari/MyRepos/finn-base/src/finn/core/modelwrapper.py", line 142, in transform
    transformed_model
  File "/home/askari/MyRepos/finn/src/finn/transformation/streamline/reorder.py", line 652, in apply
    model = model.transform(InferShapes())
  File "/home/askari/MyRepos/finn-base/src/finn/core/modelwrapper.py", line 142, in transform
    transformed_model
  File "/home/askari/MyRepos/finn-base/src/finn/transformation/infer_shapes.py", line 88, in apply
    model = ModelWrapper(si.infer_shapes(model.model))
  File "/home/askari/.local/lib/python3.6/site-packages/onnx/shape_inference.py", line 42, in infer_shapes
    inferred_model_str = C.infer_shapes(model_str, check_type, strict_mode, data_prop)
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:Transpose, node name: Transpose_3): [ShapeInferenceError] Inferred shape and existing shape differ in dimension 2: (64) vs (56)
```

**Reply by fpjentzsch** [reply]
Compatibility with Vitis HLS is not fully there yet. I suggest to use Vivado/Vivado HLS 2020.1

Some other tips:
- Did you try with the latest FINN dev branch?
- You should only need to set FINN_XILINX_PATH and FINN_XILINX_VERSION, and you should definitely not run ./run-docker.sh as root.
- FINN should be cloned to a directory where your user has full access (e.g. under your home)

**Reply by fpjentzsch** [reply]
Also, you can get rough performance estimates without access to Xilinx tools, namely the FINN-calculated resource estimation and expected cycle count per HLS block (for the applied folding configuration), from which you can derive bottlenecks in the pipeline and thus an upper bound on throughput.

The best performance "estimate" without actual deployment would be RTL simulation of the whole accelerator, which yields accurate throughput measurements. Naturally, this step requires Vivado HLS.

**Reply by hossein1387** [reply]
Thanks for the reply. To answer your questions:

- **Did you try with the latest FINN dev branch?**   Yes I am using the latest FINN dev branch.
- **You should only need to set FINN_XILINX_PATH and FINN_XILINX_VERSION, and you should definitely not run ./run-docker.sh as root.** After your suggestion, I am now back to my Mac just to get the estimates and not the rtl estimates.
- **FINN should be cloned to a directory where your user has full access (e.g. under your home)** I am running Finn in my home directory.


As I said, I am using this notebook: https://github.com/Xilinx/finn/blob/main/notebooks/end2end_example/cybersecurity/3-build-accelerator-with-finn.ipynb

However, I am unable to run `build.build_dataflow_cfg(model_file, cfg_estimates)` even for simple models. For instance, I created a simple 1 Conv layer
<img width="857" alt="Screen Shot 2021-11-17 at 10 07 26 AM" src="https://user-images.githubusercontent.com/6366236/142226188-846523ad-c7e1-4755-bd87-26ff693807d2.png">


but I get the following error:

```
Building dataflow accelerator from /Users/hossein/MyRepos/Experiments/finn/SimpleConv.onnx
Intermediate outputs will be generated in $(/Users/hossein/finn_output)
Final outputs will be generated in output_estimates_only
Build log is at output_estimates_only/build_dataflow.log
Running step: step_qonnx_to_finn [1/8]
Running step: step_tidy_up [2/8]
Running step: step_streamline [3/8]
Running step: step_convert_to_hls [4/8]
Running step: step_create_dataflow_partition [5/8]
Traceback (most recent call last):
  File "/Users/hossein/MyRepos/finn/src/finn/builder/build_dataflow.py", line 166, in build_dataflow_cfg
    model = transform_step(model, cfg)
  File "/Users/hossein/MyRepos/finn/src/finn/builder/build_dataflow_steps.py", line 317, in step_create_dataflow_partition
    assert len(sdp_nodes) == 1, "Only a single StreamingDataflowPartition supported."
AssertionError: Only a single StreamingDataflowPartition supported.
> /Users/hossein/MyRepos/finn/src/finn/builder/build_dataflow_steps.py(317)step_create_dataflow_partition()
    315     )
    316     sdp_nodes = parent_model.get_nodes_by_op_type("StreamingDataflowPartition")
--> 317     assert len(sdp_nodes) == 1, "Only a single StreamingDataflowPartition supported."
    318     sdp_node = sdp_nodes[0]
    319     sdp_node = getCustomOp(sdp_node)
```

As I mentioned before, I even test it with your own onnx models and I cannot pass `build.build_dataflow_cfg` function. Could you please let me any possible solution?

**Reply by hossein1387** [reply]
I did a simple experiment. Assuming that Brevitas can generate models compatible for FINN, I wrote the kernel above in Brevitas and export it to FINN as below:


```
from torch import nn
from torch.nn import Module
import torch.nn.functional as F
import brevitas.nn as qnn
from brevitas.export import FINNManager

class SimpleConv(Module):
    def __init__(self):
        super(SimpleConv, self).__init__()
        self.training = False
        self.conv1 = qnn.QuantConv2d(64, 64, 3, weight_bit_width=2)

    def forward(self, x):
        out = self.conv1(x)
        return out

simple_conv = SimpleConv()

FINNManager.export(simple_conv, input_shape=(1, 64, 32, 32), export_path='SimpleConv.onnx')
                                                                         
```

I used then used `SimpleConv.onnx` with this FINN-example: https://github.com/Xilinx/finn-examples/blob/main/build/resnet50/build.py


where I got the following performance estimate:

```{
  "critical_path_cycles": 0,
  "max_cycles": 1,
  "max_cycles_node_name": "",
  "estimated_throughput_fps": 250000000.0,
  "estimated_latency_ns": 0.0
}
"ConvDoublePacked_Batch_0": {
    "BRAM_18K": 43,
    "BRAM_efficiency": 0.37209302325581395,
    "LUT": 7762,
    "URAM": 0,
    "URAM_efficiency": 1,
    "DSP": 1536
  },
  "total": {
    "BRAM_18K": 43.0,
    "LUT": 7762.0,
    "URAM": 0.0,
    "DSP": 1536.0
  }
}
```


However, if i do the same thing with other models, like resnet18. resnet50 etc, this method does not work and I get an error like this:

```
Analysing models/resnet9.onnx...
Building dataflow accelerator from models/resnet9.onnx
Intermediate outputs will be generated in $(/Users/hossein/finn_output)
Final outputs will be generated in output_resnet9_U250
Build log is at output_resnet9_U250/build_dataflow.log
Running step: step_resnet50_tidy [1/6]
Running step: step_resnet50_streamline [2/6]
Traceback (most recent call last):
  File "/Users/hossein/MyRepos/finn/src/finn/builder/build_dataflow.py", line 166, in build_dataflow_cfg
    model = transform_step(model, cfg)
  File "/Users/hossein/MyRepos/finn-examples/build/resnet50/custom_steps.py", line 204, in step_resnet50_streamline
    model = step_resnet50_streamline_nonlinear(model, cfg)
  File "/Users/hossein/MyRepos/finn-examples/build/resnet50/custom_steps.py", line 195, in step_resnet50_streamline_nonlinear
    model = model.transform(trn)
  File "/Users/hossein/MyRepos/finn-base/src/finn/core/modelwrapper.py", line 141, in transform
    (transformed_model, model_was_changed) = transformation.apply(
  File "/Users/hossein/MyRepos/finn/src/finn/transformation/streamline/reorder.py", line 558, in apply
    init1 = model.get_initializer(prod1.input[1])
  File "/Users/hossein/miniforge3/lib/python3.9/site-packages/google/protobuf/internal/containers.py", line 209, in __getitem__
    return self._values[key]
IndexError: list index out of range
> /Users/hossein/miniforge3/lib/python3.9/site-packages/google/protobuf/internal/containers.py(209)__getitem__()
-> return self._values[key]

 ```
 
 @rpitonak @maltanar 
 I am still hoping someone could explain to me how I can get estimate of my models and not the ones that are provided in FINN zoo.

**Reply by maltanar** [reply]
@hossein1387 to provide estimates for a network, the FINN compiler needs to be able to first convert everything into HLS layers, and to be able to do there are a number of things that need to match up (e.g. data layout, quantization, certain other restrictions for other layers, successful streamlining). Currently, the out-of-the-box FINN flow is only able to process relatively simple linear topologies, adding custom steps is required to make it work for ResNets and similar. We hope to increase FINN's flexibility in the coming releases, but for now, making custom models (e.g. with non-linear topologies) requires some familiarity with FINN's internals for writing custom steps.

**Reply by hossein1387** [reply]
@maltanar thanks a lot for your answer.
